json-decode.go

```diff
@@ -210,46 +202,12 @@ func (cdc *Codec) decodeReflectJSONArrayOrSlice(bz []byte, info *TypeInfo, rv re
 	return nil
 }
 
-func safeNil(rv reflect.Value) bool {
```

Comment:
I feel like this is in general an "unsafe" function to have around... it seems like its primary side-effect would be masking bugs.

Reply:
Well if you have a nil value that implements json.Unmarshaler and you directly invoke json.Unmarshaler on it, that isn't a subtle bug to be masked, it's a situation that needs to be handled e.g https://play.golang.org/p/R9IUZ4C-JeQ
```go
package main

import (
	"encoding/json"
	"fmt"
	"reflect"
)

type st struct {
	Foo string
}

func (s *st) UnmarshalJSON(b []byte) error {
	s.Foo = string(b)
	return nil
}

var jsonUnmarshaler = reflect.TypeOf(new(json.Unmarshaler)).Elem()

func main() {
	in := []byte("foo-bar")
	sv := reflect.ValueOf((*st)(nil))

	var err error
	if sv.Type().Implements(jsonUnmarshaler) {
		err = sv.Interface().(json.Unmarshaler).UnmarshalJSON(in) // Without checking if it is nil, we've failed with a panic
	} else {
		err = json.Unmarshal(in, sv)
	}

	fmt.Printf("in: %+v\nerr: %v\n", sv, err)
}
```
We must be able to detect such cases and if it isn't any one of those types for which we can invoke IsNil() those are more crashes.

Reply:
It doesn't make sense to me that you wouldn't know what type sv is.

In your example, you know that sv is a `*st` which can be nil.  If you claim that you don't know what sv's type is... well, there's sv.CanAddr() which returns false if sv is nil.

Reply:
>It doesn't make sense to me that you wouldn't know what type sv is.

I don't follow, sv's Type is well known, it is `*st`

>In your example, you know that sv is a *st which can be nil. If you claim that you don't know what sv's type is...
 
I know what type sv is, I just need to know if it implements `json.Unmarshaler`, hence the `sv.Type().Implements(jsonUnmarshaler)`

> well, there's sv.CanAddr() which returns false if sv is nil.

Great option actually, thanks for this suggestion, I totally missed that.